### PR TITLE
New plugin "systemd_units" for linux

### PR DESCRIPTION
**Pull request by benschweizer**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

File: plugins/inputs/services/services_windows.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+// SampleConfig returns sample configuration options.
+func (services *Services) SampleConfig() string {
+	return `
+  ## no options
+`
+}
+
+// Gather parses wmi outputs and adds counters to the Accumulator
+func (services *Services) Gather(acc telegraf.Accumulator) error {
+	//var dst []Win_32Service
```
Comment (glinton) [inline plugins/inputs/services/services_windows.go]: Remove commented code

**Comment by danielnelson** [comment]
One big picture question on this pull request is how should we handle this change with respect to the win_services plugin.

We could either go for a unified plugin for all init systems or we could have specialized plugins for each init system.

If we decided to do a unified plugin then I think we need to provide greater unification between backends.  For example, all the backends should use the same set of state codes and the format of the state should also be shared.

If we go with specialized plugins then I think the consistency requirement can be less rigid.

File: plugins/inputs/services/services_linux.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  ## The default timeout of 1s for systemctl execution can be overridden here:
+  # timeout = "1s"
+`
+}
+
+// Gather parses systemctl outputs and adds counters to the Accumulator
+func (services *Services) Gather(acc telegraf.Accumulator) error {
+	out, e := services.systemctl(services.Timeout)
+	if e != nil {
+		acc.AddError(e)
```
[the remark this answers is not in the document]
Comment (cxcv) [reply plugins/inputs/services/services_linux.go]: ok

File: plugins/inputs/services/readme.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+> services,host=host2.example.com,name=CryptSvc state="Running",status=0i 1533731001000000000
+> services,host=host2.example.com,name=dbupdate state="Stopped",status=2i 1533731001000000000
+...
+```
+
+# Statuses
+
+| Value | Meaning  | Description                             |
+| ----- | -------  | -----------                             |
+| 0     | Ok       | Service state is withour failure        |
```
[the remark this answers is not in the document]
Comment (cxcv) [reply plugins/inputs/services/readme.md]: ok

File: plugins/inputs/services/services_windows.go (hunk 1 above)
Comment (cxcv) [reply plugins/inputs/services/services_windows.go]: ok

**Comment by cxcv** [comment]
@danielnelson @vlastahajek
Good point, I've missed the win_services plugin, and the intention is the same.
I think the code is similar, but as you stated win_services has a more complete list of states.
Generally spoken, I think a unified plugin would be better, as you can use the same alert rules.
May be it would be good to i) merge the code and ii) keep the defined states.
What do you think?

File: plugins/inputs/services/services_linux.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (glinton) [reply plugins/inputs/services/services_linux.go]: Is there a reason you're not returning the error here `return e`?

**Comment by cxcv** [comment]
@danielnelson @vlastahajek
Meanwhile I had a look into win_services.
The first difference is that win_services uses the go-native service manager while this plugin uses 3rd party wmi. The field names are different with both APIs. Next, win_services keeps more raw-data where this services plugin reduces the output to a more simple ok/warning/critical/unknown scheme. Overall, the win_services module looks more extensive and might be a fit for more use cases.
From there, renaming this module to systemd_units and dropping windows support is an option. Any comments?

**Comment by vlastahajek** [comment]
Win_services plugin was created to answer multiple requests for it and it was validated by users. As you pointed, your implementation is limited in comparing to win_services, but on the other hand it adds status field, which in not available via service manager

If functionalities are unified, it makes sense for me to have plugin that provides same interface to system services on all systems, on the other hand, users of win_services would need to switch, to new plugin, but if done properly , it may not  be a problem.

@danielnelson, WDYT?

**Comment by danielnelson** [comment]
I think what makes the most sense initially is to convert this to a systemd service only plugin.

In the future we could take a stab at making a generalized service input, which would be restricted to a subset of metrics.  Having the specialized plugins will help as we will be able to take a broad look at what data is available on all platforms.  Even with a multiplatform service input it may be useful to have ones specific to the particular init system, to expose more advanced metrics.

**Comment by cxcv** [comment]
I agree that having specific plugins for each system and a generic overall plugin would be nice.
For a pure systemd plugin, I would rename tags/fields to be systemd specific and drop the computed status. Though, what would you do differently for a generic overall services plugin? May be we are already pretty close.

**Comment by danielnelson** [comment]
It seems like the common are the service name and a general service state (stopped, running):
```
services,host=localhost,origin=localhost,name=telegraf,state=running state_code=4i
```

If we could build off of this so that systemd is a superset of these values we could extend the values:
```
services,host=localhost,origin=localhost,name=telegraf.service,state=running,status=warning state_code=4i,status_code=1i
```

We would need to make some larger changes to win_services if we wanted to fit it in.

**Comment by cxcv** [comment]
Hello,

I've set the PR to WIP, I plan to update it in the next days.
As there's a working plugin for windows, I would focus this one on systemd.
A great unified plugin would be nice, may be in a 2nd step afterwards.
Would this be ok for everyone involved?

Cheers
Benjamin

**Comment by amarao** [comment]
Hello!

It's really cool plugin, and I would be happy to see it in the telegraf, but it currently lacks one important metric: `NRestarts` (which says how many times service was restarted due to failure). Could you please add it?

P.S. Example of InfluxDB failed few times due to lack of memory:

```
systemctl show influxdb -p NRestarts
NRestarts=9
systemctl restart influxdb
systemctl show influxdb -p NRestarts
NRestarts=0
```
(as you can see it's been cleared on manual restart).

**Comment by cxcv** [comment]
@danielnelson I've updated the plugin:
- removed redundant windows services checks as they are meanwhile covered by win_services
- added support for systemd unit types beyond services
- returns all fields as reported by systemctl
If there are no further comments, you could merge this.

**Comment by cxcv** [comment]
> It's really cool plugin, and I would be happy to see it in the telegraf, but it currently lacks one important metric: `NRestarts` (which says how many times service was restarted due to failure). Could you please add it?

Nice idea, but we'd need to query each unit. Is it possible to have this in one short, e.g. as an option to list-units ?

**Comment by cxcv** [comment]
@danielnelson 
I've updated the plugin:
- removed redundant code that is meanwhile covered in the win_services plugin
- added support of unit types beyond .service
- returns all available fields (load, active, sub)

If there are no further comments, it's done and it could be merged.
Cheers
Benjamin

File: plugins/inputs/systemd_units/README.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+```
+
+### Metrics
+- systemd_units:
+  - tags:
+    - name (string, unit name)
+  - fields:
+    - loaded (int, see below)
+    - active (int, see below)
+    - running (int, see below)
```
Comment (stefanhorning) [inline plugins/inputs/systemd_units/README.md]: Shouldn't this list be `load`, `active` and `sub`?

Comment (cxcv) [reply plugins/inputs/systemd_units/README.md]: you are right thanks for spotting this

**Comment by Marlinc** [comment]
Nice work, this is a great addition to Telegraf. Will try and test this

File: plugins/inputs/systemd_units/systemd_units_linux.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+func (systemd_units *SystemdUnits) Description() string {
+	return "Gather systemd units state"
+}
+
+// SampleConfig returns sample configuration options.
+func (systemd_units *SystemdUnits) SampleConfig() string {
+	return `
+  ## The default timeout of 1s for systemctl execution can be overridden here:
+  # timeout = "1s"
+  # unittype = "service"
```
Comment (glinton) [inline plugins/inputs/systemd_units/systemd_units_linux.go]: Same comments as the readme

File: plugins/inputs/systemd_units/systemd_units_linux.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+		data := strings.Fields(line)
+		if len(data) < 4 {
+			acc.AddError(fmt.Errorf("Error parsing line (expected at least 4 fields): %s", line))
+			continue
+		}
+		tags := map[string]string{
+			"name": data[0],
+		}
+		// unit_load_state_table, see
```
Comment (glinton) [inline plugins/inputs/systemd_units/systemd_units_linux.go]: Make these state table consts and pull them out of gather.

Comment (cxcv) [reply plugins/inputs/systemd_units/systemd_units_linux.go]: I've moved them, but maps cannot be const!?

File: plugins/inputs/systemd_units/README.md [hunk 7]
```diff
@@ -10,13 +10,15 @@ This plugin is related to the [win_services module](../win_services/), which
...
 
 In addition to services, this plugin can gather other unit types as well,
-see ```systemctl list-units --all --type help``` for possible options.
+see `systemctl list-units --all --type help` for possible options.
 
 ### Configuration
 ```
 [[inputs.systemd_units]]
-  ## The default timeout of 1s for systemctl execution can be overridden here:
+  ## The default timeout of 1s for systemctl execution can be overridden
```
Comment (glinton) [inline plugins/inputs/systemd_units/README.md]: Maybe just use `## Set timeout for systemctl execution`

File: plugins/inputs/systemd_units/README.md [hunk 8]
```diff
@@ -10,13 +10,15 @@ This plugin is related to the [win_services module](../win_services/), which
...
+  ## here:
   # timeout = "1s"
+  ## Filter for a specific unit types, default is "service":
```
Comment (glinton) [inline plugins/inputs/systemd_units/README.md]: Add a newline above this line

File: plugins/inputs/systemd_units/systemd_units_linux.go [hunk 9]
```diff
@@ -24,6 +24,132 @@ type systemctl func(Timeout internal.Duration, UnitType string) (*bytes.Buffer,
 
 const measurement = "systemd_units"
 
+// unit_load_state_table, see
+// https://github.com/systemd/systemd/blob/c87700a1335f489be31cd3549927da68b5638819/src/basic/unit-def.c#L87
+var load_map = map[string]int{
+	// err: 0,
```
Comment (glinton) [inline plugins/inputs/systemd_units/systemd_units_linux.go]: Let's remove all the commented out values, if they aren't being used, it doesn't seem valuable to have lingering commented code lying around.

File: plugins/inputs/systemd_units/README.md (hunk 8 above)
Comment (glinton) [reply plugins/inputs/systemd_units/README.md]: Same for `SampleConfig`

**Comment by PsyanticY** [comment]
Hi,  any updates on this

**Pull request merged**